File: smart_akshaya_windows_application/lib/services/google_sheets_service.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:googleapis/sheets/v4.dart' as sheets;
import 'package:googleapis/drive/v3.dart' as drive;
import 'package:googleapis_auth/auth_io.dart';
import '../config/google_sheets_config.dart';

abstract class GoogleSheetsServiceBase {
  bool get isInitialized;
  Future<void> init(String credentialsJson);
  Future<String> getOrCreateSpreadsheet(String? spreadsheetId);
  Future<String> createNewSpreadsheet();
  Future<void> appendRow(String spreadsheetId, String sheetName, List<dynamic> row);
  Future<List<List<dynamic>>> getRows(String spreadsheetId, String sheetName);
  Future<void> updateRow(String spreadsheetId, String sheetName, int rowIndex, List<dynamic> row);
  Future<void> updateRowColumns(String spreadsheetId, String sheetName, int rowIndex, Map<String, dynamic> columnUpdates);
  Future<void> clearRow(String spreadsheetId, String sheetName, int rowIndex, int numColumns);
  Future<List<Map<String, dynamic>>> fetchDriveFiles(String folderId);
  Future<Uint8List> downloadDriveFile(String fileId);
}

class GoogleSheetsService implements GoogleSheetsServiceBase {
  static final GoogleSheetsService _instance = GoogleSheetsService._internal();
  factory GoogleSheetsService() => _instance;
  GoogleSheetsService._internal();

  sheets.SheetsApi? _sheetsApi;
  drive.DriveApi? _driveApi;
  bool _initialized = false;

  bool get isInitialized => _initialized;

  /// Initializes the Sheets API client with the service account credentials JSON.
  Future<void> init(String credentialsJson) async {
    if (_initialized) return;

    try {
      final Map<String, dynamic> credentialsMap = json.decode(credentialsJson);
      final accountCredentials = ServiceAccountCredentials.fromJson(credentialsMap);
      
      final client = await clientViaServiceAccount(
        accountCredentials,
        GoogleSheetsConfig.scopes,
      );

      _sheetsApi = sheets.SheetsApi(client);
      _driveApi = drive.DriveApi(client);
      _initialized = true;
      print('Google API Services successfully initialized.');
    } catch (e) {
      print('Error initializing Google Sheets Service: $e');
      _initialized = false;
      rethrow;
    }
  }

  sheets.SheetsApi get sheetsApi {
    if (!_initialized || _sheetsApi == null) {
      throw StateError('GoogleSheetsService has not been initialized. Call init() first.');
    }
    return _sheetsApi!;
  }

  drive.DriveApi get driveApi {
    if (!_initialized || _driveApi == null) {
      throw StateError('GoogleSheetsService has not been initialized. Call init() first.');
    }
    return _driveApi!;
  }

  @override
  Future<List<Map<String, dynamic>>> fetchDriveFiles(String folderId) async {
    if (folderId.isEmpty) return [];
    try {
      final query = "'$folderId' in parents and trashed=false";
      final response = await driveApi.files.list(
        q: query,
        $fields: 'files(id, name, mimeType, webViewLink, thumbnailLink)',
        pageSize: 100,
        orderBy: 'name',
      );

      final files = response.files ?? [];
      return files.map((f) => {
        'id': f.id ?? '',
        'title': f.name ?? 'Untitled Form',
        'mime_type': f.mimeType ?? '',
        'drive_link': f.webViewLink ?? '',
        'thumbnail_link': f.thumbnailLink ?? '',
      }).toList();
    } catch (e) {
      print('Error fetching drive files: $e');
      return [];
    }
  }

  @override
  Future<Uint8List> downloadDriveFile(String fileId) async {
    try {
      final response = await driveApi.files.get(
        fileId,
        downloadOptions: drive.DownloadOptions.fullMedia,
      ) as drive.Media;

      final List<int> bytes = [];
      await for (final chunk in response.stream) {
        bytes.addAll(chunk);
      }
      return Uint8List.fromList(bytes);
    } catch (e) {
      print('Error downloading drive file: $e');
      rethrow;
    }
  }

  /// Verifies if sheets "staff details" and "service details" exist, and if not, creates them.
  /// If [spreadsheetId] is null or empty, creates a new spreadsheet and initializes it.
  Future<String> getOrCreateSpreadsheet(String? spreadsheetId) async {
    if (spreadsheetId == null || spreadsheetId.trim().isEmpty) {
      return await createNewSpreadsheet();
    }

    try {
      // Attempt to retrieve spreadsheet details to check if it exists and has the required worksheets
      final spreadsheet = await sheetsApi.spreadsheets.get(spreadsheetId);
      final existingSheetNames = spreadsheet.sheets
          ?.map((s) => s.properties?.title?.toLowerCase().trim())
          .toList() ?? [];

      bool hasStaff = existingSheetNames.contains(GoogleSheetsConfig.staffSheetName.toLowerCase());
      bool hasService = existingSheetNames.contains(GoogleSheetsConfig.serviceSheetName.toLowerCase());
      bool hasCustomer = existingSheetNames.contains(GoogleSheetsConfig.customerSheetName.toLowerCase());
      bool hasExpense = existingSheetNames.contains(GoogleSheetsConfig.expenseSheetName.toLowerCase());

      List<sheets.Request> requests = [];

      if (!hasStaff) {
        requests.add(sheets.Request(addSheet: sheets.AddSheetRequest(properties: sheets.SheetProperties(title: GoogleSheetsConfig.staffSheetName))));
      }
      if (!hasService) {
        requests.add(sheets.Request(addSheet: sheets.AddSheetRequest(properties: sheets.SheetProperties(title: GoogleSheetsConfig.serviceSheetName))));
      }
      if (!hasCustomer) {
        requests.add(sheets.Request(addSheet: sheets.AddSheetRequest(properties: sheets.SheetProperties(title: GoogleSheetsConfig.customerSheetName))));
      }
      if (!hasExpense) {
        requests.add(sheets.Request(addSheet: sheets.AddSheetRequest(properties: sheets.SheetProperties(title: GoogleSheetsConfig.expenseSheetName))));
      }

      if (requests.isNotEmpty) {
        final batchRequest = sheets.BatchUpdateSpreadsheetRequest(requests: requests);
        await sheetsApi.spreadsheets.batchUpdate(batchRequest, spreadsheetId);
        print('Added missing worksheets to spreadsheet: $spreadsheetId');
        
        // Write headers for any newly added sheet
        if (!hasStaff) await _writeStaffHeaders(spreadsheetId);
        if (!hasService) await _writeServiceHeaders(spreadsheetId);
        if (!hasCustomer) await _writeCustomerHeaders(spreadsheetId);
        if (!hasExpense) await _writeExpenseHeaders(spreadsheetId);
      }

      return spreadsheetId;
    } catch (e) {
      print('Error validating spreadsheet $spreadsheetId, creating a new one: $e');
      return await createNewSpreadsheet();
    }
  }

  /// Creates a brand new Spreadsheet with correct worksheets and headers.
  Future<String> createNewSpreadsheet() async {
    try {
      final newSpreadsheet = sheets.Spreadsheet(
        properties: sheets.SpreadsheetProperties(title: 'Smart Akshaya Database'),
        sheets: [
          sheets.Sheet(properties: sheets.SheetProperties(title: GoogleSheetsConfig.staffSheetName)),
          sheets.Sheet(properties: sheets.SheetProperties(title: GoogleSheetsConfig.serviceSheetName)),
          sheets.Sheet(properties: sheets.SheetProperties(title: GoogleSheetsConfig.customerSheetName)),
          sheets.Sheet(properties: sheets.SheetProperties(title: GoogleSheetsConfig.expenseSheetName)),
        ],
      );

      final created = await sheetsApi.spreadsheets.create(newSpreadsheet);
      final newId = created.spreadsheetId;
      if (newId == null) throw Exception('Failed to retrieve spreadsheet ID from Google response');

      print('Successfully created new Google Spreadsheet with ID: $newId');
      
      // Write headers to all sheets
      await _writeStaffHeaders(newId);
      await _writeServiceHeaders(newId);
      await _writeCustomerHeaders(newId);
      await _writeExpenseHeaders(newId);

      return newId;
    } catch (e) {
      print('Error creating new Google Spreadsheet: $e');
      rethrow;
    }
  }

  Future<void> _writeStaffHeaders(String spreadsheetId) async {
    final headers = [
      ['ID', 'Name', 'Address', 'Mobile', 'Email', 'User Type', 'Status', 'Password']
    ];
    await _writeRange(spreadsheetId, '${GoogleSheetsConfig.staffSheetName}!A1', headers);
  }

  Future<void> _writeServiceHeaders(String spreadsheetId) async {
    final headers = [
      [
        'Service Name',
        'Website',
        'Department Fee',
        'Service Charge',
        'Commission',
        'Allow Edit',
        'Followup Days',
        'Default Wallet'
      ]
    ];
    await _writeRange(spreadsheetId, '${GoogleSheetsConfig.serviceSheetName}!A1', headers);
  }

  Future<void> _writeCustomerHeaders(String spreadsheetId) async {
    final headers = [['ID', 'Name', 'Mobile', 'Email', 'Address', 'Remarks', 'Total Paid', 'GPay/UPI', 'Cash', 'Balance']];
    await _writeRange(spreadsheetId, '${GoogleSheetsConfig.customerSheetName}!A1', headers);
  }

  Future<void> _writeExpenseHeaders(String spreadsheetId) async {
    final headers = [['ID', 'Date', 'Category', 'Amount', 'Description']];
    await _writeRange(spreadsheetId, '${GoogleSheetsConfig.expenseSheetName}!A1', headers);
  }

  Future<void> _writeRange(String spreadsheetId, String range, List<List<dynamic>> values) async {
    final valueRange = sheets.ValueRange.fromJson({
      'values': values,
    });
    await sheetsApi.spreadsheets.values.update(
      valueRange,
      spreadsheetId,
      range,
      valueInputOption: 'USER_ENTERED',
    );
  }

  /// Appends a single row to a sheet.
  Future<void> appendRow(String spreadsheetId, String sheetName, List<dynamic> row) async {
    final valueRange = sheets.ValueRange.fromJson({
      'values': [row],
    });
    await sheetsApi.spreadsheets.values.append(
      valueRange,
      spreadsheetId,
      '$sheetName!A1',
      valueInputOption: 'USER_ENTERED',
    );
  }

  /// Reads all rows of a sheet.
  Future<List<List<dynamic>>> getRows(String spreadsheetId, String sheetName) async {
    try {
      final response = await sheetsApi.spreadsheets.values.get(spreadsheetId, '$sheetName!A:Z');
      return response.values ?? [];
    } catch (e) {
      print('Error fetching rows from sheet $sheetName: $e');
      return [];
    }
  }

  /// Updates a specific row in a sheet (using 1-based indexing, e.g., rowIndex = 2 for second row, which is the first data row after headers).
  Future<void> updateRow(String spreadsheetId, String sheetName, int rowIndex, List<dynamic> row) async {
    final range = '$sheetName!A$rowIndex';
    await _writeRange(spreadsheetId, range, [row]);
  }

  /// Updates specific columns of a row by fetching the row first, updating columns, and writing back.
  Future<void> updateRowColumns(String spreadsheetId, String sheetName, int rowIndex, Map<String, dynamic> columnUpdates) async {
    final rows = await getRows(spreadsheetId, sheetName);
    if (rows.isEmpty) throw Exception('Sheet is empty');
    final headers = rows[0].map((e) => e.toString().trim().toLowerCase()).toList();

    final List<dynamic> currentRow = List.from(rows[rowIndex - 1]);
    while (currentRow.length < headers.length) {
      currentRow.add('');
    }

    columnUpdates.forEach((key, val) {
      final idx = headers.indexOf(key.toLowerCase());
      if (idx != -1) {
        while (currentRow.length <= idx) {
          currentRow.add('');
        }
        currentRow[idx] = val;
      }
    });

    await updateRow(spreadsheetId, sheetName, rowIndex, currentRow);
  }

  /// Helper to delete/clear a row or set status. In Google Sheets, deleting rows can be done via batch update.
  /// For simplicity here, we can set Status column to 'Inactive' or overwrite a row.
  Future<void> clearRow(String spreadsheetId, String sheetName, int rowIndex, int numColumns) async {
    final range = '$sheetName!A$rowIndex:${_getColumnLetter(numColumns)}$rowIndex';
    final emptyRow = List.filled(numColumns, '');
    await _writeRange(spreadsheetId, range, [emptyRow]);
  }

  String _getColumnLetter(int colCount) {
    if (colCount <= 0) return 'A';
    String result = '';
    int count = colCount;
    while (count > 0) {
      int modulo = (count - 1) % 26;
      result = String.fromCharCode(65 + modulo) + result;
      count = ((count - modulo) / 26).toInt();
    }
    return result;
  }
}

File: smart_akshaya_web_application/src/services/googleSheetsService.js
import { getAccessToken } from './googleSheetsAuth';
import { getSpreadsheetId } from '../config/sheetsConfig';

/**
 * Reads all rows from a Google Sheet.
 * @param {string} sheetName - Name of the sheet (tab) to read from.
 * @returns {Promise<Array<Array<any>>>}
 */
export async function getRows(sheetName) {
  try {
    console.log(`[getRows] Initiating fetch for sheet: ${sheetName}`);
    const spreadsheetId = getSpreadsheetId();
    console.log(`[getRows] Spreadsheet ID: ${spreadsheetId}`);
    
    const token = await getAccessToken();
    console.log(`[getRows] Token acquired (length: ${token?.length})`);

    const range = `'${sheetName}'!A:Z`;
    const url = `https://sheets.googleapis.com/v4/spreadsheets/${spreadsheetId}/values/${encodeURIComponent(range)}`;
    console.log(`[getRows] Request URL: ${url}`);

    const response = await fetch(url, {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${token}`,
      },
      cache: 'no-store',
    });

    console.log(`[getRows] Response Status: ${response.status} ${response.statusText}`);

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`[getRows] Fetch failed! Status: ${response.status}, Error: ${errorText}`);
      throw new Error(`Error fetching sheet rows: ${response.statusText} - ${errorText}`);
    }

    const data = await response.json();
    console.log(`[getRows] Success! Fetched ${data.values ? data.values.length : 0} rows from ${sheetName}`);
    return data.values || [];
  } catch (error) {
    console.error(`[getRows] Caught exception for sheet ${sheetName}:`, error);
    throw error;
  }
}

/**
 * Appends a row to a Google Sheet.
 * @param {string} sheetName - Name of the sheet (tab) to append to.
 * @param {Array<any>} row - The array of row cell values.
 */
export async function appendRow(sheetName, row) {
  const spreadsheetId = getSpreadsheetId();
  const token = await getAccessToken();
  const range = `'${sheetName}'!A1`;
  const url = `https://sheets.googleapis.com/v4/spreadsheets/${spreadsheetId}/values/${encodeURIComponent(range)}:append?valueInputOption=USER_ENTERED`;

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      values: [row],
    }),
  });

  if (!response.ok) {
    throw new Error(`Error appending row: ${response.statusText}`);
  }

  return await response.json();
}

/**
 * Appends multiple rows to a Google Sheet.
 * @param {string} sheetName - Name of the sheet (tab) to append to.
 * @param {Array<Array<any>>} rows - The array of row cell values (2D array).
 */
export async function appendRows(sheetName, rows) {
  const spreadsheetId = getSpreadsheetId();
  const token = await getAccessToken();
  const range = `'${sheetName}'!A1`;
  const url = `https://sheets.googleapis.com/v4/spreadsheets/${spreadsheetId}/values/${encodeURIComponent(range)}:append?valueInputOption=USER_ENTERED`;

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      values: rows,
    }),
  });

  if (!response.ok) {
    throw new Error(`Error appending rows: ${response.statusText}`);
  }

  return await response.json();
}

/**
 * Updates a specific row in a Google Sheet.
 * @param {string} sheetName - Name of the sheet (tab) to update.
 * @param {number} rowIndex - 1-indexed row number (e.g. 2 for first data row).
 * @param {Array<any>} row - The array of new cell values.
 */
export async function updateRow(sheetName, rowIndex, row) {
  const spreadsheetId = getSpreadsheetId();
  const token = await getAccessToken();
  const range = `'${sheetName}'!A${rowIndex}`;
  const url = `https://sheets.googleapis.com/v4/spreadsheets/${spreadsheetId}/values/${encodeURIComponent(range)}?valueInputOption=USER_ENTERED`;

  const response = await fetch(url, {
    method: 'PUT',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      values: [row],
    }),
  });

  if (!response.ok) {
    throw new Error(`Error updating row: ${response.statusText}`);
  }

  return await response.json();
}

/**
 * Updates specific columns in a row by first reading the current row,
 * merging new values, then writing back.
 * @param {string} sheetName
 * @param {number} rowIndex - 1-indexed
 * @param {Object} columnUpdates - { 'column header name': newValue, ... }
 */
export async function updateRowColumns(sheetName, rowIndex, columnUpdates) {
  // Read header row to find column indexes
  const rows = await getRows(sheetName);
  if (!rows || rows.length < 1) throw new Error('Sheet is empty');
  const headers = rows[0].map((h) => h.trim().toLowerCase());

  // Get current data row (rowIndex is 1-based; rows array is 0-based)
  const currentRow = [...(rows[rowIndex - 1] || [])];

  // Apply updates
  Object.entries(columnUpdates).forEach(([key, val]) => {
    const idx = headers.indexOf(key.toLowerCase());
    if (idx !== -1) currentRow[idx] = String(val);
  });

  return updateRow(sheetName, rowIndex, currentRow);
}

/**
 * Clears the content of a row (sets values to empty strings).
 * @param {string} sheetName
 * @param {number} rowIndex - 1-indexed
 * @param {number} numColumns
 */
export async function clearRow(sheetName, rowIndex, numColumns) {
  const spreadsheetId = getSpreadsheetId();
  const token = await getAccessToken();
  
  const getColumnLetter = (colCount) => {
    let result = '';
    let count = colCount;
    while (count > 0) {
      let modulo = (count - 1) % 26;
      result = String.fromCharCode(65 + modulo) + result;
      count = Math.floor((count - modulo) / 26);
    }
    return result || 'A';
  };

  const endColumnLetter = getColumnLetter(numColumns);
  const range = `'${sheetName}'!A${rowIndex}:${endColumnLetter}${rowIndex}`;
  const url = `https://sheets.googleapis.com/v4/spreadsheets/${spreadsheetId}/values/${encodeURIComponent(range)}?valueInputOption=USER_ENTERED`;

  const emptyRow = Array(numColumns).fill('');

  const response = await fetch(url, {
    method: 'PUT',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      values: [emptyRow],
    }),
  });

  if (!response.ok) {
    throw new Error(`Error clearing row: ${response.statusText}`);
  }

  return await response.json();
}

/**
 * Deletes a row from a Google Sheet.
 * @param {string} sheetName - Name of the sheet (tab) to delete from.
 * @param {number} rowIndex - The 1-indexed row number to delete.
 */
export async function deleteRow(sheetName, rowIndex) {
  const spreadsheetId = getSpreadsheetId();
  const token = await getAccessToken();

  // First, we need to get the sheet ID (gid) for the sheetName
  const sheetMetaUrl = `https://sheets.googleapis.com/v4/spreadsheets/${spreadsheetId}`;
  const metaResponse = await fetch(sheetMetaUrl, {
    headers: { Authorization: `Bearer ${token}` }
  });

  if (!metaResponse.ok) {
    throw new Error(`Failed to fetch spreadsheet metadata: ${metaResponse.statusText}`);
  }

  const metaData = await metaResponse.json();
  const sheet = metaData.sheets.find(s => s.properties.title === sheetName);
  
  if (!sheet) {
    throw new Error(`Sheet ${sheetName} not found`);
  }

  const sheetId = sheet.properties.sheetId;

  // Use batchUpdate to delete the row
  const url = `https://sheets.googleapis.com/v4/spreadsheets/${spreadsheetId}:batchUpdate`;
  
  const response = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      requests: [
        {
          deleteDimension: {
            range: {
              sheetId: sheetId,
              dimension: 'ROWS',
              startIndex: rowIndex - 1, // 0-indexed, inclusive
              endIndex: rowIndex        // 0-indexed, exclusive
            }
          }
        }
      ]
    })
  });

  if (!response.ok) {
    const errorText = await response.text();
    console.error(`[deleteRow] failed: ${response.status} - ${errorText}`);
    throw new Error(`Error deleting row: ${response.statusText}. Details: ${errorText}`);
  }

  return await response.json();
}

/**
 * Logs a wallet transaction to the 'Wallet Transactions' sheet.
 * @param {string} walletName 
 * @param {string} type 
 * @param {number} amount 
 * @param {number} closingBalance 
 * @param {string} description 
 */
export async function logWalletTransaction(walletName, type, amount, closingBalance, description = '', staffName = '', billId = '') {
  try {
    const now = new Date();
    const day = String(now.getDate()).padStart(2, '0');
    const month = String(now.getMonth() + 1).padStart(2, '0');
    const year = now.getFullYear();
    const dateStr = `${day}-${month}-${year}`; // Format: DD-MM-YYYY
    const timeStr = now.toLocaleTimeString('en-IN', { hour12: true });

    const row = [
      dateStr,
      timeStr,
      walletName,
      type,
      amount.toFixed(2),
      closingBalance.toFixed(2),
      description,
      staffName,
      billId
    ];

    try {
      await appendRow('Transaction History', row);
    } catch (e) {
      await appendRow('Wallet Transactions', row);
    }
  } catch (err) {
    console.error('Failed to log wallet transaction. Make sure Transaction History or Wallet Transactions sheet exists.', err);
  }
}
